### ISSUE TITLE
Support for `username_file` configuration in prometheus http basic auth config

### ISSUE THREAD

**Issue by wasim-nihal** [prometheus/prometheus#12576]
### Proposal

**Feature Request**

**Description:**
Currently, a user has to explicitly type in the sensitive data in the configuration yaml. For example, in case of basic_auth, the user needs to configure the username and password. In our organization username is also considered as sensitive data and configuring it in this way is not acceptable. 

In grafana, there is an option(called as [File Provider](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#:~:text=Ini-,File%20provider,-file%20reads%20a))  to provide a path to a file for an value using a placeholder like `$__file{<path_to_file>}`. At the runtime, the content of the file is read and substituted for the variable dynamically.

**Advantages**

1. In Kubernetes environment especially, the sensitive data can be made as secrets and mounted on the pod as files. And the placeholders in prometheus configuration can be dynamically updated with the contents of the respective files and then application is started.
2. Any field in the configuration yaml can be treated as sensitive data by the organizations. Gives them better flexibility.

**Example Usage in configuration**
 ```
- job_name: 'health_checks'
    scrape_interval: 4m
    scrape_timeout: 25s
    static_configs:
        - targets: ['organization.com']
    metrics_path: "/api/health"
    basic_auth:
      username: $__file{/etc/secret/username}
      password: $__file{/etc/secret/password}
```

**Contribution**

I am working on the changes for this feature and if you would allow, I would be happy to contribute this back to the Prometheus community.

**Comment by roidelapluie** [comment]
We already support reading passwords via file.

```
- job_name: 'health_checks'
   scrape_interval: 4m
   scrape_timeout: 25s
   static_configs:
       - targets: ['organization.com']
   metrics_path: "/api/health"
   basic_auth:
     username: username
     password_file: /etc/secret/password
```

**Comment by wasim-nihal** [comment]
Thanks for your response.  This functionality is limited to passwords right? In some organizations, including mine, even username is treated as sensitive data and the value cannot be provided directly in the configuration yaml. So, with this feature, a user can create a Kubernetes secret for username and mount it to the pod and can further the configuration yaml placeholder(`$__file{secret_mount_file}`) can be substituted with the content of the username secret.

With this, it will provide flexibility for the users/organizations on what fields they want to treat as sensitive data based on their requirements.

**Comment by roidelapluie** [comment]
I am open to adding a username_file if you think it's useful.

Please note that Prometheus is not just substituting a field with the content of a file. When a file is specified, we read it on every request, enabling rolling changes of passwords without configuration reloads.

**Comment by wasim-nihal** [comment]
yes, having `username_file`, similar to `password_file` would definitely be useful for some of the organizations like ours. 
@roidelapluie , if this requirement (introduction of `username_file`) is okay, I would like to contribute here.

**Comment by wasim-nihal** [comment]
Hi @roidelapluie , please find the PR for having `username_file` configuration (file based basic auth username configuration).

- https://github.com/prometheus/common/pull/511
- https://github.com/prometheus/prometheus/pull/12749

**Comment by roypeter** [comment]
Hi @wasim-nihal, thank you for incorporating this feature; it has proven to be quite useful to us. Might you have an expected release date for it?

**Comment by beorn7** [comment]
Hello from the bug scrub!

This has been released as prometheus/common v0.45. Which made it into Prometheus with v2.49 via a dependency update. We missed documenting the change in the CHANGELOG, and more importantly did not update the config documentation. This only happened with v3. Sorry for that. But now it is definitely out there.
